### PR TITLE
fix: re-enable custom chain iconUrl and iconBackground

### LINKED ISSUE #1527: [bug] custom chain iconUrl and iconBackground not working.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### RainbowKit Version

1.0.11

### wagmi Version

1.4.2

### Current Behavior

When configuring chains, there's an option to pass in custom iconUrl and iconBackground as shown below:

```
import { Chain } from '@rainbow-me/rainbowkit'
import { mainnet, optimism } from 'wagmi/chains';

const defaultChains: Chain[] = [
  {
    ...mainnet,
    iconUrl: 'https://example.com/icons/ethereum.png',
    iconBackground: 'transparent'
  },
  {
    ...optimism,
    iconUrl: 'https://example.com/icons/optimism.png',
    iconBackground: 'transparent'
  },
];

const { chains } = configureChains(defaultChains, [
  alchemyProvider({ apiKey: process.env.ALCHEMY_ID }),
  publicProvider(),
]);
```

But this customisation no longer takes effect due to [this recent change](https://github.com/rainbow-me/rainbowkit/commit/f1e98e8401d0a92dda1bf9f660f2ba630708f5d0#:~:text=/**%20%40description%20Decorates,\)\)%3B) on this merged pull request #1408 by @DanielSinclair 

Moving the `chainMetadataById` spread to come after the configured wagmi chain data spread overwrites whatever customisation that was previously possible.

### Expected Behavior

I would suggest since the goal for #1408 is to ensure colloquial chain names, I would suggest that only the `name` property is put after the provided chain spread as shown below:

```
export const provideRainbowKitChains = <Chain extends RainbowKitChain>(
  chains: Chain[],
): Chain[] =>
  chains.map(chain => {
    const chainMeta = chainMetadataById[chain.id] ?? {};

    return {
      ...chainMeta,
      ...chain,
      name: chainMeta.name ?? chain.name,
    }
  })
```

This way, `chainId` and `name` cannot be overwritten while maintaining `iconUrl` and `iconBackground` customisation.

### Steps To Reproduce

_No response_

### Link to Minimal Reproducible Example (CodeSandbox, StackBlitz, etc.)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by kadetXx**
Allow custom chain iconUrl and backgroundUrl while maintaining favor for RainbowKitChain name overrrides.

As shown in the screenshot below, `ConnectButtonRenderer` and `ChainModal` still favors `RainbowKitChain` name overrrides without disabling custom icons for each chain. 

Screenshot 2023-09-25 at 6 20 27 PM

<br />
<br />

To achieve this state in the screenshot for **temporary** testing purposes, I mapped through the chains array passed into `configureChains` and then passed in a custom iconUrl (eth logo)  and also a custom name (Nameee) just to validate that preference for `RainbowKitChain` name is still maintained after re-enabling custom chain icons.

### CODE REVIEW

**Comment by kadetXx** [comment]
This PR solves issue #1527

**Comment by DanielSinclair** [comment]
@kadetXx Thanks for spotting this and the PR! Performed some refactoring and added test coverage of this scenario. Will merge today

File: packages/rainbowkit/test/index.tsx [hunk 1]
```diff
@@ -24,7 +24,7 @@ export function renderWithProviders(
     props?: Omit<RainbowKitProviderProps, 'children'>;
   },
 ) {
-  const supportedChains = options?.chains || defaultChains;
+  const supportedChains: Chain[] = options?.chains || defaultChains;
```
Comment (KosmosKey) [inline packages/rainbowkit/test/index.tsx:27]: Great type check ✅

File: packages/rainbowkit/src/components/RainbowKitProvider/provideRainbowKitChains.ts [hunk 2]
```diff
@@ -128,7 +128,12 @@ const chainMetadataById = Object.fromEntries(
...
-  }));
+  chains.map((chain) => {
+    const defaultMetadata = chainMetadataById[chain.id] ?? {};
+    return {
+      ...chain,
+      name: defaultMetadata.name ?? chain.name, // favor colloquial names
+      iconUrl: chain.iconUrl ?? defaultMetadata.iconUrl,
+      iconBackground: chain.iconBackground ?? defaultMetadata.iconBackground,
+    } as Chain;
+  });
```
Comment (KosmosKey) [inline packages/rainbowkit/src/components/RainbowKitProvider/provideRainbowKitChains.ts:139]: ✨

**Review by 0xmkh — APPROVED** [review]
Code looks good, no bugs found and all tests are passing. Ready to merge ✅

**Comment by kadetXx** [comment]
> @kadetXx Thanks for spotting this and the PR! Performed some refactoring and added test coverage of this scenario. Will merge today

Happy to help!

**Pull request merged**